### PR TITLE
Clean up Flutter app launching and debugger hook

### PR DESCRIPTION
**Pull request by skybrian**
- Introduced Launcher class as a central place for app launching code
- Split up run configuration parameters into SdkFields and BazelFields
- Cleaned up usage of ObservatoryConnector. Use FlutterApp in more places.
- Added tests for run configuration persistence.

### CODE REVIEW

**Comment by skybrian** [comment]
Sorry it's so big; the changes are somewhat related. I recommend reading Launcher first, then SdkFields and BazelFields, then ObservatoryConnector, and that will give an idea of what's going on.

File: src/io/flutter/run/SdkFields.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2016 The Chromium Authors. All rights reserved.
```
Comment (pq) [inline src/io/flutter/run/SdkFields.java]: EDIT: feel free to ignore.  Just realizing that these are all moved files.

File: src/io/flutter/run/Launcher.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+ * <p>Normally creates a debugging session, which is needed for hot reload.
+ */
+public class Launcher extends CommandLineState {
+  private final @NotNull VirtualFile workDir;
+
+  /**
+   * The file or directory holding the Flutter app's source code.
+   * This determines how the analysis server resolves URI's (for breakpoints, etc).
+   *
+   * <p>If a file, this should be the file containing the main() method.
```
Comment (skybrian) [inline src/io/flutter/run/Launcher.java]: I haven't dug into the debugger code enough to know for sure, but my guess is that if it's the wrong directory entirely, the debugger will have trouble translating between package URL's and source filenames, and therefore setting breakpoints won't work.

File: src/io/flutter/run/SdkFields.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    final DartSdk sdk = DartPlugin.getDartSdk(project);
+    if (sdk == null) {
+      throw new RuntimeConfigurationError(FlutterBundle.message("dart.sdk.is.not.configured"),
+                                          () -> DartConfigurable.openDartSettings(project));
+    }
+    return sdk;
+  }
+
+  static @NotNull VirtualFile checkLaunchFile(String path) throws RuntimeConfigurationError {
+    // TODO(skybrian) also check that it's a Flutter app?
```
Comment (skybrian) [inline src/io/flutter/run/SdkFields.java]: Yep

File: third_party/intellij-plugins-dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceDebugProcessZ.java [hunk 4]
```diff
@@ -291,9 +284,14 @@ private void connect() throws IOException {
     myVmConnected = true;
     getSession().rebuildViews();
 
-    FlutterViewMessages.sendDebugActive(getSession().getProject(), getConnector(), myVmServiceWrapper, vmService);
```
Comment (skybrian) [inline third_party/intellij-plugins-dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceDebugProcessZ.java]: Moved to superclass. See: FlutterDebugProcess.onVmConnected().

File: src/io/flutter/run/bazel/BazelRunConfig.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+import io.flutter.run.daemon.FlutterAppService;
+import io.flutter.run.daemon.RunMode;
+import org.jdom.Element;
+import org.jetbrains.annotations.NotNull;
+
+public class BazelRunConfig extends RunConfigurationBase
+  implements RunConfigurationWithSuppressedDefaultRunAction, Launcher.RunConfig {
+  private @NotNull BazelFields fields = new BazelFields();
+
+  public BazelRunConfig(final Project project, final ConfigurationFactory factory, final String name) {
```
Comment (skybrian) [inline src/io/flutter/run/bazel/BazelRunConfig.java]: Done

**Pull request merged**